README.md
Mobile-Factory-Erya


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 29/12/2020
  Features:
    - Full rewrite of the Script for increased performances
    - The Snow/Ice should now be more reactive
    - Added the Show/Hide Debug Option (Can be set inside the Mobile Factory Option GUI - System)
---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 29/12/2020
  Features:
    - Mobile Factory is not longer a mandatory Dependency, the Mod require now MF: Base
  Changes:
    - The Erya collector not longer requires an Energy Source to work
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 12/12/2020
  Features:
    - Update to Factorio 1.1
---------------------------------------------------------------------------------------------------
Version: 0.1.8
Date: 24/11/2020
  Features:
    - Allows the Mobile Factory mod to shows a Chunk temperature inside the Main GUI
---------------------------------------------------------------------------------------------------
Version: 0.1.7
Date: 16/08/2020
  Features:
    - Updated to Factorio 1.0
---------------------------------------------------------------------------------------------------
Version: 0.1.6
Date: 13/08/2020
  Bugfixes:
    - Fixed a bug where Erya Simple can't be placed inside the Dimensional Lab (Second attempt)
---------------------------------------------------------------------------------------------------
Version: 0.1.5
Date: 31/07/2020
  Bugfixes:
    - Fixed a bug where Erya Simple can't be placed inside the Dimensional Lab
---------------------------------------------------------------------------------------------------
Version: 0.1.4
Date: 30/07/2020
  Bugfixes:
    - Fixed a crash when a Surface was deleted
---------------------------------------------------------------------------------------------------
Version: 0.1.3
Date: 28/07/2020
  Optimisations:
    - Some performances optimization done
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 27/07/2020
  Bugfixes:
    - Fixed a compatibility issue with some Mods
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 27/07/2020
  Features:
    - Improved snow transitions with the other Tiles
    - Added Ice: water will now freeze
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 25/07/2020
  Features:
    - Created the "Mobile Factory: Erya" Extension

data.lua
require("prototypes/erya/erya-intermediates.lua")
require("prototypes/erya/erya-collector.lua")
require("prototypes/erya/erya-structures.lua")
require("prototypes/winter/snow-tiles.lua")

data:extend{
	{
		type="item-group",
		name="Erya",
		icon="__Mobile_Factory-Erya__/graphics/Erya/EryaPowder.png",
		icon_size="256",
		order="y"
	}
}

data:extend{
	{
		type="item-subgroup",
		name="EryaRessources",
		group="Erya",
		order="a"
	}
}

data:extend{
	{
		type="item-subgroup",
		name="EryaIntermediates",
		group="Erya",
		order="b"
	}
}

data:extend{
	{
		type="item-subgroup",
		name="EryaLogistic",
		group="Erya",
		order="c"
	}
}

data:extend{
	{
		type="item-subgroup",
		name="EryaProduction",
		group="Erya",
		order="d"
	}
}

data:extend{
	{
		type="item-subgroup",
		name="EryaWar",
		group="Erya",
		order="e"
	}
}

data:extend{
	{
		type="recipe-category",
		name="EryaPowder",
		order="d"
	}
}

migrations/0.3.0to0.4.0.lua
-- Recreate the Chunks Table and the 3D Chunk Table --
global.chunksTable = {}
global.chunksTable3D = {}
-- Scan all surfaces to find all Erya Entities --
for _, surface in pairs(game.surfaces) do
    -- Get all Erya Entity of the Surface --
    local ents = surface.find_entities_filtered{name=frozenStructures}
    -- Save all Entities --
    for _, ent in pairs(ents) do
        -- Check the Entity --
        if ent ~= nil and ent.valid == true then
            -- Get the Chunk Position --
            local cx = math.floor(ent.position.x / chunkSize)
            local cy = math.floor(ent.position.y / chunkSize)
            -- Create the Area --
            local x1 = cx * chunkSize
            local y1 = cy * chunkSize
            local x2 = x1 + chunkSize
            local y2 = y1 + chunkSize
            local area = {left_top = {x=x1, y=y1}, right_bottom = {x=x2, y=y2}}
            -- Save the Entity --
            saveChunk(cx, cy, area, ent.surface.index)
        end
    end
end

prototypes/winter/snow-tiles.lua
-- Snow Tile 1 --
local st1E = table.deepcopy(data.raw.tile["grass-1"])
st1E.name = "MFSnowTile1"
st1E.autoplace = nil
st1E.map_color = {1,1,1}
st1E.walking_speed_modifier = 0.85
st1E.variants = tile_variations_template(
	"__Mobile_Factory-Erya__/graphics/winter/snow-tile-1.png", "__base__/graphics/terrain/masks/transition-3.png",
	"__Mobile_Factory-Erya__/graphics/winter/hr-snow-tile-1.png", "__base__/graphics/terrain/masks/hr-transition-3.png",
	{
	  max_size = 4,
	  [1] = { weights = {0.085, 0.085, 0.085, 0.085, 0.087, 0.085, 0.065, 0.085, 0.045, 0.045, 0.045, 0.045, 0.005, 0.025, 0.045, 0.045 } },
	  [2] = { probability = 0.91, weights = {0.150, 0.150, 0.150, 0.150, 0.018, 0.020, 0.015, 0.025, 0.015, 0.020, 0.025, 0.015, 0.025, 0.025, 0.010, 0.025 }, },
	  [4] = { probability = 0.91, weights = {0.100, 0.80, 0.80, 0.100, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01 }, },
	  --[8] = { probability = 1.00, weights = {0.090, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.025, 0.125, 0.005, 0.010, 0.100, 0.100, 0.010, 0.020, 0.020} }
	}
  )
data:extend{st1E}

-- Ice Tile 1 --
local it1E = {} -- table.deepcopy(data.raw.tile.water)
it1E.name = "MFIceTile1"
it1E.type = "tile"
it1E.autoplace = nil
it1E.draw_in_water_layer = true
it1E.layer = 3
it1E.map_color={r=0.0941, g=0.149, b=0.066}
it1E.pollution_absorption_per_second = 0
it1E.transitions = data.raw.tile.water.transitions
it1E.walking_speed_modifier = 0.8
it1E.collision_mask =
{
  "water-tile",
  "item-layer",
  "resource-layer",
  -- "player-layer",
  "doodad-layer"
}
it1E.variants =
    {
      main =
      {
        {
			picture = "__Mobile_Factory-Erya__/graphics/winter/hr-ice1.png",
            count = 1,
            scale = 0.5,
            size = 1
        },
        {
			picture = "__Mobile_Factory-Erya__/graphics/winter/hr-ice2.png",
            count = 1,
            scale = 0.5,
            size = 2
        },
        {
			picture = "__Mobile_Factory-Erya__/graphics/winter/hr-ice3.png",
            count = 1,
            scale = 0.5,
            size = 4
        }
      },
      empty_transitions = true,
	}
data:extend{it1E}

prototypes/erya/erya-collector.lua
---------------------------------- ERYA COLLECTOR ---------------------------

-- Entity --
ecE = {}
ecE.type = "assembling-machine"
ecE.name = "EryaCollector"
ecE.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaCollectorI.png"
ecE.icon_size = 256
ecE.flags = {"placeable-neutral", "player-creation", "not-rotatable"}
ecE.minable = {mining_time = 0.2, result = "EryaCollector"}
ecE.max_health = 100
ecE.corpse = "small-remnants"
ecE.collision_box = {{-2.5, -2.35}, {2.5, 2.35}}
ecE.selection_box = ecE.collision_box
ecE.vehicle_impact_sound =  { filename = "__base__/sound/car-wood-impact.ogg", volume = 1.0 }
ecE.energy_usage = "1W"
ecE.crafting_speed = 1
ecE.crafting_categories = {"EryaPowder"}
ecE.collision_mask = {"ground-tile"}
ecE.fixed_recipe = "EryaPowder2"
ecE.energy_source =
{
    type="void",
    render_no_power_icon = false,
    render_no_network_icon = false

}
ecE.animation =
{
    layers =
    {
        {
        frame_count = 60,
        filename = "__Mobile_Factory-Erya__/graphics/Erya/EryaCollectorE.png",
        size = 600,
        line_length = 10,
        animation_speed = 0.5,
        scale = 0.35
        },
        {
            frame_count = 1,
            line_length = 1,
            repeat_count = 60,
            filename = "__Mobile_Factory-Erya__/graphics/Erya/EryaCollectorS.png",
            size = 600,
            scale = 0.35,
            shift = {0.3, 0.2},
            draw_as_shadow = true
        }
    }
}
data:extend{ecE}

-- Item --
local ecI = {}
ecI.type = "item"
ecI.name = "EryaCollector"
ecI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaCollectorI.png"
ecI.icon_size = 256
ecI.place_result = "EryaCollector"
ecI.subgroup = "EryaRessources"
ecI.order = "a"
ecI.stack_size = 10
data:extend{ecI}

-- Recipe --
local ecR = {}
ecR.type = "recipe"
ecR.name = "EryaCollector"
ecR.energy_required = 2
ecR.enabled = false
ecR.ingredients =
    {
      {"DimensionalCircuit", 10},
      {"MachineFrame2", 2}
    }
ecR.result = "EryaCollector"
data:extend{ecR}

prototypes/erya/erya-intermediates.lua
-- Erya Powder Item --
local epI = {}
epI.type = "item"
epI.name = "EryaPowder"
epI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaPowder.png"
epI.icon_size = 256
epI.subgroup = "EryaRessources"
epI.order = "b"
epI.stack_size = 1000
data:extend{epI}

-- Erya Powder Recipe --
local epR = {}
epR.type = "recipe"
epR.name = "EryaPowder"
epR.energy_required = 1
epR.enabled = false
epR.ingredients =
    {
      {"EryaSample", 3}
    }
epR.result = "EryaPowder"
epR.result_count = 1
data:extend{epR}

-- Erya Powder Recipe 2 --
local epR2 = {}
epR2.type = "recipe"
epR2.name = "EryaPowder2"
epR2.energy_required = 2
epR2.enabled = false
epR2.category = "EryaPowder"
epR2.ingredients = {}
epR2.hidden = true
epR2.result = "EryaPowder"
epR2.result_count = 1
data:extend{epR2}

-- Erya Sample Item --
local esI = {}
esI.type = "tool"
esI.name = "EryaSample"
esI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaSample.png"
esI.icon_size = 256
esI.subgroup = "EryaRessources"
esI.durability = 1
esI.infinite = false
esI.order = "c"
esI.stack_size = 1000
data:extend{esI}

-- Erya Sample Recipe --
local esR = {}
esR.type = "recipe"
esR.name = "EryaSample"
esR.energy_required = 1
esR.enabled = false
esR.ingredients =
    {
      {"EryaPowder", 1}
    }
esR.result = "EryaSample"
esR.result_count = 3
data:extend{esR}

-- Erya Plate Item --
local eplI = {}
eplI.type = "item"
eplI.name = "EryaPlate"
eplI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaPlate.png"
eplI.icon_size = 256
eplI.subgroup = "EryaIntermediates"
eplI.order = "a"
eplI.stack_size = 1000
data:extend{eplI}

-- Erya Plate Recipe --
local eplR = {}
eplR.type = "recipe"
eplR.name = "EryaPlate"
eplR.energy_required = 1
eplR.enabled = false
eplR.ingredients =
    {
        {"EryaPowder", 4}
    }
eplR.result = "EryaPlate"
data:extend{eplR}

-- Erya Wire Item --
local ewI = {}
ewI.type = "item"
ewI.name = "EryaWire"
ewI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaWire.png"
ewI.icon_size = 64
ewI.subgroup = "EryaIntermediates"
ewI.order = "c"
ewI.stack_size = 1000
data:extend{ewI}

-- Erya Wire Recipe --
local ewR = {}
ewR.type = "recipe"
ewR.name = "EryaWire"
ewR.energy_required = 1
ewR.enabled = false
ewR.ingredients =
    {
        {"EryaPlate", 1}
    }
ewR.result = "EryaWire"
data:extend{ewR}

-- Erya Circuit Item --
local ecI = {}
ecI.type = "item"
ecI.name = "EryaCircuit"
ecI.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaCircuit.png"
ecI.icon_size = 64
ecI.subgroup = "EryaIntermediates"
ecI.order = "d"
ecI.stack_size = 200
data:extend{ecI}

-- Erya Circuit Recipe --
local ecR = {}
ecR.type = "recipe"
ecR.name = "EryaCircuit"
ecR.energy_required = 1
ecR.enabled = false
ecR.ingredients =
    {
        {"EryaPlate", 1},
        {"EryaWire", 2}
    }
ecR.result = "EryaCircuit"
data:extend{ecR}

-- Erya Machine Frame 1 Item --
local emf1I = {}
emf1I.type = "item"
emf1I.name = "EryaMachineFrame1"
emf1I.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaMachineFrame1.png"
emf1I.icon_size = 256
emf1I.subgroup = "EryaIntermediates"
emf1I.order = "e"
emf1I.stack_size = 20
data:extend{emf1I}

-- Erya Machine Frame Recipe --
local emf1R = {}
emf1R.type = "recipe"
emf1R.name = "EryaMachineFrame1"
emf1R.energy_required = 1
emf1R.enabled = false
emf1R.ingredients =
    {
      {"EryaPlate", 3}
    }
emf1R.result = "EryaMachineFrame1"
data:extend{emf1R}

-- Technologie --
local eT = {}
eT.name = "Erya"
eT.type = "technology"
eT.icon = "__Mobile_Factory-Erya__/graphics/Erya/EryaPowder.png"
eT.icon_size = 256
eT.unit = {
	count=300,
	time=2,
	ingredients={
		{"DimensionalSample", 1}
	}
}
eT.prerequisites = {"DimensionalOre"}
eT.effects =
{
    {type="unlock-recipe", recipe="EryaPowder"},
    {type="unlock-recipe", recipe="EryaSample"},
    {type="unlock-recipe", recipe="EryaPlate"},
    {type="unlock-recipe", recipe="EryaWire"},
    {type="unlock-recipe", recipe="EryaCircuit"},
    {type="unlock-recipe", recipe="EryaMachineFrame1"},
    {type="unlock-recipe", recipe="EryaCollector"}
}
data:extend{eT}

control.lua
pcall(require,'__debugadapter__/debugadapter.lua')

-- How many Tick between each Chunk Update --
chunkUpdateTime = 300

-- The max Chunks checked per tick --
maxTileCheck = 100

-- The max Chunks updated per tick --
maxChunkUpdate = 15

-- The size of a Chunk --
chunkSize = 4

-- The initial temperature --
initialTemperature = 20 -- °C --

-- The maximum natural temperature --
maxNaturalTemperature = 20 -- °C --

-- The minimal temperature --
minTemperature = -273 -- °C --

-- How many degrees is added per Update --
addedTemperature = 0.25

-- At under which Temperature du frost spread --
spreadTemperature = -10

-- Higher is this number, slower the temperature will spread --
spreadFactor = 5


-- Erya Structures that freeze the environment --
frozenStructures = {
                        "EryaLamp",
                        "EryaChest1",
                        "EryaTank1",
                        "EryaBelt1",
                        "EryaItemMover",
                        "EryaUndergroundBelt1",
                        "EryaUndergroundBelt2",
                        "EryaSplitter1",
                        "EryaSplitter2",
                        "EryaLoader1",
                        "EryaLoader2",
                        "EryaInserter1",
                        "EryaMiningDrill1",
                        "EryaPumpjack1",
                        "EryaAssemblingMachine1",
                        "EryaPipe1",
                        "EryaPipeToGround1",
                        "EryaPump1",
                        "EryaWall1",
                        "EryaGate1",
                        "EryaRadar1",
                        "EryaFurnace1",
                        "EryaRefinery1",
                        "EryaChemicalPlant1"
                    }

-- List of Tiles that must be replaced by Ice --
waterTiles = {}
waterTiles["deepwater"] = true
waterTiles["deepwater-green"] = true
waterTiles["water"] = true
waterTiles["water-green"] = true
waterTiles["water-shallow"] = true
waterTiles["water-mud"] = true
waterTiles["MFIceTile1"] = true

-- The amount of temperature that a Erya structure decrease (in °C) --
heatCost = {}
heatCost["EryaLamp"] = 5
heatCost["EryaChest1"] = 3
heatCost["EryaTank1"] = 5
heatCost["EryaBelt1"] = 1
heatCost["EryaItemMover"] = 2
heatCost["EryaUndergroundBelt1"] = 1.5
heatCost["EryaUndergroundBelt2"] = 1.5
heatCost["EryaSplitter1"] = 2
heatCost["EryaSplitter2"] = 3
heatCost["EryaLoader1"] = 2
heatCost["EryaLoader2"] = 3
heatCost["EryaInserter1"] = 5
heatCost["EryaMiningDrill1"] = 10
heatCost["EryaPumpjack1"] = 10
heatCost["EryaAssemblingMachine1"] = 10
heatCost["EryaPipe1"] = 1
heatCost["EryaPipeToGround1"] = 1.5
heatCost["EryaPump1"] = 3
heatCost["EryaWall1"] = 1
heatCost["EryaGate1"] = 3
heatCost["EryaRadar1"] = 15
heatCost["EryaFurnace1"] = 12
heatCost["EryaRefinery1"] = 15
heatCost["EryaChemicalPlant1"] = 15

-- When the Mod load for the first time --
function onInit()
    
    -- Create the Chunk Table --
    global.chunksTable = {}
    global.chunksTable3D = {}

    -- The Update Index to know where the Chunk Table have to be updated --
    global.chunkTableIndex = 0

    -- Show the Debug --
    global.showDebug = false

end

-- Called every Tick --
function onTick()

    -- Check the Chunk Table Index --
    local size = table_size(global.chunksTable)
    global.chunkTableIndex = global.chunkTableIndex or 1
    if global.chunkTableIndex >= size then global.chunkTableIndex = 0 end

    -- game.print(table_size(global.chunksTable))

    -- Update the Chunk Table --
    local chunkUpdated = 0
    local maxChunkUpdateCalculated = math.min(maxChunkUpdate, math.floor(size/chunkUpdateTime)+1)
    for _=1, math.min(maxTileCheck, size) do

        -- Get the Chunk --
        local _, chunk = next(global.chunksTable, global.chunkTableIndex ~= 0 and global.chunkTableIndex or nil)

        -- Check the Chunk --
        if chunk == nil then break end

        -- Check if the Chunk has to be updated --
        if (game.tick - (chunk.lastCheck or 0)) > chunkUpdateTime then

            -- Save the Last Update --
            chunk.lastCheck = game.tick

            -- Check the Surface --
            local surface = game.get_surface(chunk.surfaceIndex)
            if surface == nil or surface.valid == false then
                removeChunk(chunk, global.chunkTableIndex+1)
            else
                -- Update the Chunk --
                updateChunk(chunk, global.chunkTableIndex+1)
                chunkUpdated = chunkUpdated + 1
                if chunkUpdated > maxChunkUpdateCalculated then break end
            end
            
        end

        -- Increase the Index --
        global.chunkTableIndex = global.chunkTableIndex + 1

    end

end

function updateChunk(chunk, key)

    -- Decrease the Temperature for every Erya Tech present --
    local eryaEnts = game.get_surface(chunk.surfaceIndex).find_entities_filtered{area=chunk.area, name=frozenStructures}
    if table_size(eryaEnts) > 0 then
        for _, ent in pairs(eryaEnts) do
            chunk.temp = math.max(chunk.temp - (heatCost[ent.name] or 0), minTemperature)
        end
    end

    -- Increase the Temperature --
    chunk.temp = chunk.temp + addedTemperature

    -- Check if the Chunk has to be removed --
    if chunk.temp >= maxNaturalTemperature then
        -- Check if there are no Frozen Tiles left --
        if table_size(game.get_surface(chunk.surfaceIndex).find_tiles_filtered{area=chunk.area, name = "MFSnowTile1"} or {}) <= 0 then
            if table_size(game.get_surface(chunk.surfaceIndex).find_tiles_filtered{area=chunk.area, name = "MFIceTile1"} or {}) <= 0 then
                removeChunk(chunk, key)
            end
        end
    end

    -- Spread the Temperature if Needed --
    if chunk.temp < spreadTemperature then
        spreadFrost(chunk)
    end

    -- Update the Tiles --
    updateTiles(chunk)

    -- Draw the Temperature Rectangle --
    if global.showDebug ~= true then return end
    rendering.draw_rectangle{color={255,0,0}, width=5, left_top=chunk.area.left_top, right_bottom=chunk.area.right_bottom, surface=game.get_surface(chunk.surfaceIndex), time_to_live=20}
    if chunk.temp > 10 then
        rendering.draw_rectangle{color={0,200,0,150}, filled=true, left_top=chunk.area.left_top, right_bottom=chunk.area.right_bottom, surface=game.get_surface(chunk.surfaceIndex), time_to_live=chunkUpdateTime+30}
    elseif chunk.temp > -150 then
        rendering.draw_rectangle{color={0,12,255,150}, filled=true, left_top=chunk.area.left_top, right_bottom=chunk.area.right_bottom, surface=game.get_surface(chunk.surfaceIndex), time_to_live=chunkUpdateTime+30}
    else
        rendering.draw_rectangle{color={87,140,255,150}, filled=true, left_top=chunk.area.left_top, right_bottom=chunk.area.right_bottom, surface=game.get_surface(chunk.surfaceIndex), time_to_live=chunkUpdateTime+30}
    end

end

-- Spread the Frost to the 8 surrounding Chunks --
function spreadFrost(chunk)

    -- Get a random X and Y to spread --
    local rX = math.random(-1, 1)
    local rY = math.random(-1, 1)

    -- Get the random Chunk --
    local iChunk = {x=chunk.x + rX, y=chunk.y + rY}

    -- Check if the Chunk already exist inside the Chunk Table --
    local table1 = global.chunksTable3D[chunk.surfaceIndex]
    if table1 ~= nil then
        local table2 = table1[iChunk.x]
        if table2 ~= nil then
            local table3 = table2[iChunk.y]
            if table3 ~= nil then
                if table3.temp > (minTemperature+(chunk.temp/spreadFactor)) and chunk.temp < table3.temp then
                    -- Update the Temperature --
                    table3.temp = math.max(table3.temp + chunk.temp/spreadFactor, minTemperature)
                    chunk.temp = chunk.temp - chunk.temp/spreadFactor
                end
                return
            end
        end
    end

    -- Create the Area --
    local x1 = iChunk.x * chunkSize
    local y1 = iChunk.y * chunkSize
    local x2 = x1 + chunkSize
    local y2 = y1 + chunkSize
    local area = {left_top = {x=x1, y=y1}, right_bottom = {x=x2, y=y2}}

    -- Save the Chunk inside the Chunks Tables --
    local savedChunk = saveChunk(iChunk.x, iChunk.y, area, chunk.surfaceIndex)
    savedChunk.temp = math.max(initialTemperature + chunk.temp/10)

end

-- Update the Chunk Tiles --
function updateTiles(chunkObj)

    -- Get random Locations --
    local x = math.random(1,chunkSize)
    local y = math.random(1,chunkSize)
    x = x + chunkObj.area.left_top.x - 1
    y = y + chunkObj.area.left_top.y - 1

    -- Get the Tile --
    local tile = game.get_surface(chunkObj.surfaceIndex).get_tile(x,y)

    -- Check the Tile --
    if tile == nil or tile.valid == false then return end

    -- Check if Snow/Ice have to be created --
    if chunkObj.temp < 0 and tile.name ~= "MFSnowTile1" and tile.name ~= "MFIceTile1" then
        -- Save the Old Tile --
        tile.surface.set_hidden_tile(tile.position, tile.name)
        -- Create Snow/Ice Tiles --
        local tileName = waterTiles[tile.name] == nil and "MFSnowTile1" or "MFIceTile1"
        tile.surface.set_tiles({{name=tileName, position=tile.position}})
    elseif chunkObj.temp > 0 and (tile.name == "MFSnowTile1" or tile.name == "MFIceTile1") then
        if tile.hidden_tile ~= nil and game.tile_prototypes[tile.hidden_tile] ~= nil then
            tile.surface.set_tiles({{name=tile.hidden_tile, position=tile.position}})
        else
            tile.surface.set_tiles({{name="dirt-7", position=tile.position}})
        end
    end
    
end

-- Called when Something was placed --
function somethingWasPlaced(event)

    -- Get the Entity --
    local ent = event.created_entity or event.entity or event.destination

    -- Check the Entity --
    if ent == nil or ent.valid == false then return end

    -- Check if this is a Erya Tech --
    if heatCost[ent.name] == nil then return end

    -- Get the Chunk Position --
    local cx = math.floor(ent.position.x / chunkSize)
    local cy = math.floor(ent.position.y / chunkSize)

    -- Create the Area --
    local x1 = cx * chunkSize
    local y1 = cy * chunkSize
    local x2 = x1 + chunkSize
    local y2 = y1 + chunkSize
    local area = {left_top = {x=x1, y=y1}, right_bottom = {x=x2, y=y2}}

    -- Check if the Chunk does not already exist --
    if global.chunksTable3D[ent.surface.index] ~= nil and global.chunksTable3D[ent.surface.index][cx] ~= nil and global.chunksTable3D[ent.surface.index][cx][cy] ~= nil then
        -- The chunk exist, do nothing --
    else
        -- Save the Chunk inside the Chunks Tables --
        saveChunk(cx, cy, area, ent.surface.index)
    end

end

-- Save a Chunk inside the Chunks Tables --
function saveChunk(posX, posY, area, surfaceIndex)
    -- Create the Chunk Table --
    local chunkTable = {surfaceIndex=surfaceIndex, x=posX, y=posY, area=area, temp=initialTemperature, lastCheck=nil}
    -- Save the Chunk inside the 3D Chunk Table --
    global.chunksTable3D[surfaceIndex] = global.chunksTable3D[surfaceIndex] or {}
    global.chunksTable3D[surfaceIndex][posX] = global.chunksTable3D[surfaceIndex][posX] or {}
    global.chunksTable3D[surfaceIndex][posX][posY] = global.chunksTable3D[surfaceIndex][posX][posY] or chunkTable
    -- Save the Chunk inside the Chunk Table --
    table.insert(global.chunksTable, global.chunksTable3D[surfaceIndex][posX][posY])
    -- Return the chunk --
    return global.chunksTable3D[surfaceIndex][posX][posY]
end

-- Remove a Chunk from the Chunks Tables --
function removeChunk(chunk, key)

    -- Try to Remove the Chunk from the 3D Table --
    if global.chunksTable3D[chunk.surfaceIndex] ~= nil and global.chunksTable3D[chunk.surfaceIndex][chunk.x] ~= nil and global.chunksTable3D[chunk.surfaceIndex][chunk.x][chunk.y] ~= nil then
        global.chunksTable3D[chunk.surfaceIndex][chunk.x][chunk.y] = nil
        if table_size(global.chunksTable3D[chunk.surfaceIndex][chunk.x]) <= 0 then global.chunksTable3D[chunk.surfaceIndex][chunk.x] = nil end
        if table_size(global.chunksTable3D[chunk.surfaceIndex]) <= 0 then global.chunksTable3D[chunk.surfaceIndex] = nil end
    end

    -- Remove the Chunk from Chunk Table --
    table.remove(global.chunksTable, key)

end

-- Event --
script.on_init(onInit)
script.on_event(defines.events.on_tick, onTick)
script.on_event(defines.events.on_built_entity, somethingWasPlaced)
script.on_event(defines.events.script_raised_built, somethingWasPlaced)
script.on_event(defines.events.script_raised_revive, somethingWasPlaced)
script.on_event(defines.events.on_robot_built_entity, somethingWasPlaced)
script.on_event(defines.events.on_entity_cloned, somethingWasPlaced)


--------------------------- Interface ---------------------------

-- Return the Chunk Temperature --
function getChunkTemp(surfaceIndex, posX, posY)
    local x = math.floor(posX / chunkSize)
    local y = math.floor(posY / chunkSize)
    if global.chunksTable3D == nil or global.chunksTable3D[surfaceIndex] == nil or global.chunksTable3D[surfaceIndex][x] == nil or global.chunksTable3D[surfaceIndex][x][y] == nil or global.chunksTable3D[surfaceIndex][x][y].temp == nil then
        return 20
    end
    return global.chunksTable3D[surfaceIndex][x][y].temp
end

-- Show/Hide the Debug Information --
function showDebug(set)
    global.showDebug = set
end

-- Create the Interface --
remote.add_interface("EryaCom", {
    getChunkTemp=getChunkTemp,
    showDebug=showDebug
})